packages/@glimmer/runtime/lib/vm/rehydrate-builder.ts

```diff
@@ -156,30 +185,45 @@ export class RehydrateBuilder extends NewElementBuilder implements ElementBuilde
     this.blockDepth--;
 
     let { candidate } = currentCursor;
-    // rehydrating
+
+    let isRehydrating = false;
+
     if (candidate !== null) {
+      isRehydrating = true;
       assert(
         openBlockDepth === this.blockDepth,
         'when rehydrating, openBlockDepth should match this.blockDepth here'
       );
-      if (isComment(candidate) && getCloseBlockDepth(candidate) === openBlockDepth) {
-        currentCursor.candidate = this.remove(candidate);
+
+      if (isCloseBlock(candidate) && getBlockDepth(candidate) === openBlockDepth) {
+        let nextSibling = this.remove(candidate);
+        this.candidate = nextSibling;
         currentCursor.openBlockDepth--;
       } else {
+        // close the block and clear mismatch in parent container
+        // we will be either at the end of the element
+        // or at the end of our containing block
         this.clearMismatch(candidate);
+        isRehydrating = false;
       }
-      // if the openBlockDepth matches the blockDepth we just closed to
-      // then restore rehydration
     }
-    if (currentCursor.openBlockDepth === this.blockDepth) {
-      assert(
-        currentCursor.nextSibling !== null &&
-          isComment(currentCursor.nextSibling) &&
-          getCloseBlockDepth(currentCursor.nextSibling) === openBlockDepth,
-        'expected close block to match rehydrated open block'
-      );
-      currentCursor.candidate = this.remove(currentCursor.nextSibling!);
-      currentCursor.openBlockDepth--;
+
+    if (!isRehydrating) {
```

Comment:
For clarity, it might be better if this did an equality check

```js
if (isRehydrating === false) {
```

Reply:
Good idea, done!